Support bfloat16 and complex dtypes for logical_not

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43537 Support bfloat16 and complex dtypes for logical_not**
* #43174 Add complex number support for binary logical operators

Differential Revision: [D23751950](https://our.internmc.facebook.com/intern/diff/D23751950)

xuhdev committed:
Support bfloat16 and complex dtypes for logical_not

[ghstack-poisoned]

xuhdev committed:
Update on "Support bfloat16 and complex dtypes for logical_not"




[ghstack-poisoned]

mruberry commented on test/test_torch.py:
```diff
@@ -6266,39 +6266,32 @@ def test_bitwise_xor(self, device):
-        a_np = np.array(data, dtype=torch_to_numpy_dtype_dict[dtype])
-        self.assertEqual(np.logical_not(a_np), torch.logical_not(a).to('cpu'))
-        self.assertEqual(np.logical_not(a_np, out=a_np), a.logical_not_().to('cpu'))
+        if dtype == torch.bfloat16 or dtype.is_complex:  # numpy doesn't support these dtypes
```
NumPy does support logical not on complex types, doesn't it? 

```
a = np.array((1 + 1j))
np.logical_not(a)
: False
```

mruberry replied:
For bfloat16 you can set an explicit dtype of float32 with a pattern like:

```
if dtype is torch.bfloat16:
  np_dtype = np.float32
else:
  np_dtype = torch_to_numpy_dtype_dict[dtype]
```

xuhdev replied:
Sorry for the late reply. I've removed complex from this line.

I've tried your suggestion but `self.assertEqual(a, a.cpu().numpy())` (a few lines below) always leads to `TypeError` for bfloat16.

I have also rebased the stack.

mruberry commented on test/test_torch.py:
```diff
@@ -6266,39 +6266,32 @@ def test_bitwise_xor(self, device):
-        np.logical_not(a.cpu().numpy(), out=out_np)
-        self.assertEqual(out_np, out.to('cpu'))
+        out = torch.empty_like(a, dtype=out_dtype, device=device)
+        if torch.bfloat16 in dtypes or dtype.is_complex or out_dtype.is_complex:  # numpy doesn't support these dtypes
```
Same question here about complex NumPy support

mruberry commented:
Hey @xuhdev, sorry this went so long without a response. I've approved the other 3 PRs in this stack but have a question about this one.

xuhdev committed:
Update on "Support bfloat16 and complex dtypes for logical_not"




[ghstack-poisoned]

mruberry reviewed (approved):
Nice cleanup! Thanks @xuhdev. Now I have to get this stack merged.

xuhdev committed:
Update on "Support bfloat16 and complex dtypes for logical_not"


Differential Revision: [D23751950](https://our.internmc.facebook.com/intern/diff/D23751950)

[ghstack-poisoned]

mruberry commented:
Congrats, @xuhdev! Finally the entire stack is landed!

xuhdev commented:
@mruberry Finally! Thanks for all the landing work!
